miasm2/core/asmblock.py

```diff
@@ -359,132 +323,209 @@ class AsmSymbolPool(object):
-        if label is None:
-            label = self.add_label(offset, offset)
-        return label
-
-    def rename_label(self, label, newname):
-        """Rename the @label name to @newname"""
-        if newname in self._name2label:
+        """Get a loc_key from its @offset, create it if it doesn't exist"""
+        loc_key = self.getby_offset(offset)
+        if loc_key is None:
+            loc_key = self.add_location(offset)
+        return loc_key
+
+    def rename_location(self, loc_key, newname):
+        """Rename the @loc_key name to @newname"""
+        if newname in self._name_to_loc_key:
             raise ValueError('Symbol already known')
-        self._name2label.pop(label.name, None)
-        label.name = newname
-        self._name2label[label.name] = label
+        name = self._loc_key_to_name[loc_key]
+        assert name is not None
+        _, self._name_to_loc_key.pop(name)
+        self._loc_key_to_name[loc_key] = newname
 
-    def set_offset(self, label, offset):
-        """Pin the @label from at @offset
+    def set_offset(self, loc_key, offset):
+        """Pin the @loc_key from at @offset
         Note that there is a special case when the offset is a list
         it happens when offsets are recomputed in resolve_symbol*
         """
-        if label is None:
-            raise ValueError('label should not be None')
-        if not label.name in self._name2label:
-            raise ValueError('label %s not in symbol pool' % label)
-        if offset is not None and offset in self._offset2label:
-            raise ValueError('Conflict in label %s' % label)
-        self._offset2label.pop(label.offset, None)
-        label.offset = offset
-        if is_int(label.offset):
-            self._offset2label[label.offset] = label
+        assert isinstance(loc_key, LocKey)
+        assert offset not in self._offset_to_loc_key
+        if loc_key not in self._loc_keys:
+            raise ValueError('Foreign loc_key %s' % loc_key)
+
+        old_offset = self._loc_key_to_offset.pop(loc_key, None)
+        _ = self._offset_to_loc_key.pop(old_offset, None)
+
+        self._loc_key_to_offset[loc_key] = offset
+        self._offset_to_loc_key[offset] = loc_key
 
     @property
-    def labels(self):
-        """Return all labels"""
-        return self._labels
+    def locations(self):
+        """Return all loc_keys"""
+        return self._loc_keys
 
     @property
     def items(self):
-        """Return all labels"""
-        warnings.warn('DEPRECATION WARNING: use "labels" instead of "items"')
-        return list(self._labels)
+        """Return all loc_keys"""
+        warnings.warn('DEPRECATION WARNING: use "locations" instead of "items"')
+        return list(self._loc_keys)
 
 
     def __str__(self):
-        return reduce(lambda x, y: x + str(y) + '\n', self._labels, "")
+        return reduce(lambda x, y: x + str(y) + '\n', self._loc_keys, "")
 
     def __getitem__(self, item):
```

Comment:
Not sure if this API is used, and if it use, it has an ambiguity: why not basing it on `loc_key`?

Reply:
done